WordPress/Jetpack/Classes/ViewRelated/WordPress Migration/Types/MigrationFlowCoordinator.swift

```diff
@@ -0,0 +1,72 @@
+import UIKit
+
+final class MigrationFlowCoordinator {
+
+    // MARK: - Properties
+
+    private let account: WPAccount
+
+    private(set) lazy var navigationController: UINavigationController = {
+        let rootViewController = self.viewController(for: currentStep)
+        return MigrationFlowNavigationController(rootViewController: rootViewController)
+    }()
+
+    // MARK: - State
+
+    private var currentStep = MigrationStep.welcome {
+        didSet {
+            self.didUpdateCurrentStep(currentStep)
+        }
+    }
+
+    // MARK: - Init
+
+    init() {
+        let context = ContextManager.sharedInstance().mainContext
+        guard let account = try? WPAccount.lookupDefaultWordPressComAccount(in: context) else {
+            fatalError()
+        }
+        self.account = account
+    }
+
+    // MARK: - Step Updates
+
+    private func didUpdateCurrentStep(_ newValue: MigrationStep) {
+        let viewController = self.viewController(for: newValue)
+        self.navigationController.setViewControllers([viewController], animated: true)
+    }
+
+    // MARK: - User Interaction
+
+    private func didTapContinue(in step: MigrationStep) {
+        if let nextStep = Self.nextStep(from: step) {
+            self.currentStep = nextStep
+        } else {
+            // Migration Flow is done
+        }
+    }
+
+    // MARK: - View Controllers
+
+    private func viewController(for step: MigrationStep) -> UIViewController {
+        switch step {
+        case .welcome:
+            let viewModel = MigrationWelcomeViewModel(account: account, primaryAction: { self.didTapContinue(in: step) })
```

Comment:
adding one more consideration here, from the comment above, where I suggested the injection of the view model: because we might need different view models depending on the step, maybe we can inject a view model factory instead?

Reply:
So what I had in mind for this is either we inject an instance of the base `WelcomeStepViewModel` or an instance of its subclass. For example, let's say we add the `Notification` migration step. We could do the following:
```swift
final class MigrationNotificationViewModel: MigrationStepViewModel {
    // ...
}

private func viewController(for step: MigrationStep) -> UIViewController {
      // ...
      case .notification:
              let viewModel = MigrationNotificationViewModel(params)
              return MigrationNotificationViewControlle(viewModel)
}
```
But I'm curious to learn more about injecting a view model factory instead, what do you mean by that?

Reply:
The general idea of a `ViewControllerFactory` or a `ViewModelFactory` is that an object (the factory) takes care of the specific implementation. For example, that object could have a method with this signature

`func makeMigrationNotificationViewController() -> UIViewController`

and we can consume it in here. What that method does, what viewModel injects in the vc and what other properties are needed to it, does not concern the coordinator, whose sole responsibility is to instantiate and push the right view depending on it's state.
Does it make sense?

Reply:
👋 I can see both sides here. If `MigrationFlowCoordinator` were more focused, perhaps it wouldn't need an injected `account` property and that could be up to the factory.

On the other hand it seems we're confident this project will have a limited scope and relatively short lifetime, so unless refactoring boosts testability I'm personally fine either way.